pgeocode.py

```diff
@@ -311,9 +311,61 @@ def query_postal_code(self, codes):
             response = response.iloc[0]
         return response
 
-    def query_location(self, name):
-        """Get locations information from a community/minicipality name"""
-        pass
+    def query_location(self, name: str, top_k: int = 500, fuzzy_threshold: int = 80,
+                       col: str = "place_name") -> pd.DataFrame:
+        """Get location information from a place name
+
+        Parameters
+        ----------
+        name: str
+          string containing place names to search for
+        top_k: int
+          maximum number of results (rows in DataFrame) to return
+        fuzzy_threshold: int
+          threshold (lower bound) for fuzzy string search for finding the place name
+        col: str
+          which column in the internal data to search through
+
+        Returns
+        -------
+        df : pandas.DataFrame
+          a pandas.DataFrame with the relevant information for all matching place names (or empty if no match was found)
+        """
+        contains_matches = self._str_contains_search(name, col)
+        if len(contains_matches) > 0:
+            return contains_matches.iloc[:top_k]
+
+        fuzzy_matches = self._fuzzy_search(name, col, threshold=fuzzy_threshold)
+        if len(fuzzy_matches) > 0:
+            return fuzzy_matches.iloc[:top_k]
+
+        return pd.DataFrame(columns=self._data.columns)
+
+    def _str_contains_search(self, text: str, col: str) -> pd.DataFrame:
+        candidates = self._data[col]
+        text_len = len(text)
+        match_mask = candidates.str.lower().str.contains(text.lower())
+        if match_mask.sum() == 0:
+            return pd.DataFrame(columns=self._data.columns)
+
+        length_diffs = self._data[col][match_mask].apply(lambda x: len(x) - text_len)
+        min_diff = length_diffs.min()
+        return self._data[match_mask][length_diffs == min_diff]
+
+    def _fuzzy_search(self, text: str, col: str, threshold: float = 80) -> pd.DataFrame:
+        try:
+            # thefuzz is not required to install pgeocode,
+            # it is an optional dependency for enabling fuzzy search
+            from thefuzz import fuzz
+        except ImportError:
+            return pd.DataFrame(columns=self._data.columns)
+
+        fuzzy_scores = self._data[col].apply(lambda x: fuzz.ratio(str(x), text))
+        max_score = np.max(fuzzy_scores)
+        if max_score >= threshold:
+            return self._data[fuzzy_scores == max_score]
+        else:
+            return pd.DataFrame(columns=self._data.columns)
```

Comment:
I don't really understand what is happening with the max_score there. I would have expected this to be something like,
```
```suggestion
        mask = fuzzy_scores >= threshold
        return self._data[mask]
```
which should also work when no row matches. Though I haven't looked at thefuzz package in detail.

Otherwise threshold is never used.

Reply:
The idea was that as a user, you would like to obtain the best matching results according to the scores output by `thefuzz`, not just all results above the threshold. But we can change it of course, then the end user is responsible for further filtering the output dataframe if there are many matches above the fuzzy threshold (or just setting a higher threshold).

Reply:
Adapted it as you suggested